lms/djangoapps/dashboard/sysadmin.py

```diff
@@ -480,6 +485,173 @@ def post(self, request):
+                    one_seq_verts = get_children(sequential)
+                    for vert in one_seq_verts:
+                        verticals.append(vert)
+
+                # getting vertical (unit) names
+                unit_names = []
+                for vertical in verticals:
+                    unit_name = course_ordered[vertical].get('display_name')
+                    unit_names.append(unit_name)
+
+                header = [_('username'), _('email'), _('registration date'), _('enrolled courses'), ('last login'), 'last visit', ]
+
+                # preparing list for counting visits for course units (last raw in the table)
+                visit_count = ['', '', '', '', '', 'Visits:', ]
+
+                for name in unit_names:
+                    header.append(name)  # adding unit names to the table header
+                    visit_count.append(0)
+
+                for u in enrolled_students:  # u - User object, enrolled to current course
+                    db_query = models.StudentModule.objects.filter(
+                        course_id__exact=course_key,
+                        student_id=u.id
+                    )   # querySet StudentModule
+                    last_visit_id = db_query.all().order_by('-modified').values_list('module_state_key', flat=True)[:1]
+                    try:
+                        last_visit = course_ordered[last_visit_id[0]].get('display_name')
+                    except Exception:
+                        last_visit = 'New to course'
+
+                    visited_in_course = []
+                    for block in db_query.all().values('module_state_key'):
+                        msk = block['module_state_key']
+                        visited_in_course.append(msk)  # usageKeys for course parts where student have been
+
+                    # getting other student enrollments
+                    enrollments = list(get_course_enrollments(u, course_org_filter, org_filter_out_set))
+                    enrolled_ids = ', '.join([enrollment.course_id.course for enrollment in enrollments])
+
+                    # Comparing course structure chunks (problem, video, html) with units visited in course
+                    # Need to get positions for sequentials because this is the
+                    # only one method to check whether student visit html-only lesson or no
+                    seq_positions = {}  # getting positions for sequential module_types
+                    visited_in_seq = db_query.filter(module_type='sequential')
+                    for v in visited_in_seq.values('module_state_key', 'state'):
+                        # structure example: v = {
+                        # 'module_state_key': u'block-v1:edX+DemoX+Demo_Course+type@sequential+block@edx_introduction',
+                        # 'state': u'{"position": 1}'
+                        # }
+                        key = v['module_state_key']
+                        val = json.loads(v['state']).get('position')
```

Comment:
this part is very error prone, you should either use {}.get(key, default) or wrap it in try-catch and catch the key error and json error in case v['state'] contains some garbage. This is a general recommendation - if you are dealing with data structures that you did not create yourself, but got from somewhere, you can not be too careful, anything can be in there.